### ISSUE TITLE
Crash at the end of the battle

### ISSUE THREAD

**Issue by gestern** [vcmi/vcmi#2487]
**Describe the bug**
The game crashes **at the end** of the battle (before receiving bonuses). See call stack for more details:
See: https://github.com/vcmi/vcmi/blob/ad8695ac91a71193c7cd423cdc6d735c0803daf8/lib/bonuses/CBonusSystemNode.cpp#L92

**Game logs**

### Call Stack
```
Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libvcmi.dylib                 	       0x1051b1d9c BonusList::push_back(std::__1::shared_ptr<Bonus> const&) + 28
1   libvcmi.dylib                 	       0x1051b1f38 BonusList::getAllBonuses(BonusList&) const + 44
2   libvcmi.dylib                 	       0x1051bf704 CBonusSystemNode::getAllBonusesRec(BonusList&, CSelector const&) const + 284
3   libvcmi.dylib                 	       0x1051bf6b0 CBonusSystemNode::getAllBonusesRec(BonusList&, CSelector const&) const + 200
4   libvcmi.dylib                 	       0x1051bfb34 CBonusSystemNode::getAllBonuses(CSelector const&, CSelector const&, CBonusSystemNode const*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) const + 260
5   libvcmi.dylib                 	       0x1051c6ff4 IBonusBearer::hasBonus(CSelector const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) const + 72
6   libvcmi.dylib                 	       0x10574e624 AFactionMember::getNativeTerrain() const + 120
7   libvcmi.dylib                 	       0x105319174 CGHeroInstance::getNativeTerrain() const + 68
8   libvcmi.dylib                 	       0x105411650 TurnInfo::TurnInfo(CGHeroInstance const*, int) + 424
9   libvcmi.dylib                 	       0x105403ac4 CPathfinderHelper::updateTurnInfo(int) + 100
10  libvcmi.dylib                 	       0x105404fec CPathfinderHelper::CPathfinderHelper(CGameState*, CGHeroInstance const*, PathfinderOptions const&) + 172
11  libvcmi.dylib                 	       0x10540bf00 SingleHeroPathfinderConfig::SingleHeroPathfinderConfig(CPathsInfo&, CGameState*, CGHeroInstance const*) + 312
12  libvcmi.dylib                 	       0x105229578 CGameState::calculatePaths(CGHeroInstance const*, CPathsInfo&) + 76
13  vcmiclient                    	       0x102487584 CClient::getPathsInfo(CGHeroInstance const*) + 228
14  vcmiclient                    	       0x10211086c AdventureMapInterface::onTileHovered(int3 const&) + 1100
15  vcmiclient                    	       0x10225d3a0 non-virtual thunk to MapViewActions::mouseMoved(Point const&, Point const&) + 108
16  vcmiclient                    	       0x1021d3210 EventDispatcher::dispatchMouseMoved(Point const&, Point const&) + 564
17  vcmiclient                    	       0x1021c83a8 std::__1::__function::__func<CGuiHandler::fakeMouseMove()::$_0, std::__1::allocator<CGuiHandler::fakeMouseMove()::$_0>, void ()>::operator()() + 52
18  vcmiclient                    	       0x1021bbd74 InputHandler::fetchEvents() + 64
19  vcmiclient                    	       0x1023e9794 main + 17376
20  dyld                          	       0x187e6ff28 start + 2236
```

### `VCMI_server.txt`
```
[2023-Aug-05 20:37:04.885534] DEBUG global - Hero hero.core.nagash.name visits Грифоны (54:4)
[2023-Aug-05 20:37:04.887941] DEBUG global - Round -1
[2023-Aug-05 20:37:23.865815] DEBUG global - Round 0
[2023-Aug-05 20:37:36.175680] DEBUG global - Round 1
[2023-Aug-05 20:37:42.510215] DEBUG global - Calculating casualties for purple battle stack [0]: 33 of creatures.core.skeleton.name.plural from slot 0 of armyobj=4008
[2023-Aug-05 20:37:42.510261] DEBUG global - Count: 33; base count: 33
[2023-Aug-05 20:37:42.510272] DEBUG global - Calculating casualties for purple battle stack [1]: 35 of creatures.core.walkingDead.name.plural from slot 1 of armyobj=4008
[2023-Aug-05 20:37:42.510279] DEBUG global - Count: 35; base count: 35
[2023-Aug-05 20:37:42.510287] DEBUG global - Calculating casualties for purple battle stack [2]: 4 of creatures.core.blackKnight.name.plural from slot 2 of armyobj=4008
[2023-Aug-05 20:37:42.510293] DEBUG global - Count: 4; base count: 4
[2023-Aug-05 20:37:42.510300] DEBUG global - Calculating casualties for purple battle stack [3]: 2 of creatures.core.lich.name.plural from slot 3 of armyobj=4008
[2023-Aug-05 20:37:42.510306] DEBUG global - Count: 2; base count: 2
[2023-Aug-05 20:37:42.510312] DEBUG global - Calculating casualties for purple battle stack [4]: 25 of creatures.core.vampireLord.name.plural from slot 4 of armyobj=4008
[2023-Aug-05 20:37:42.510318] DEBUG global - Count: 25; base count: 25
[2023-Aug-05 20:37:42.510324] DEBUG global - Calculating casualties for unnamed battle stack [5]: 0 of creatures.core.griffin.name.plural from slot 0 of armyobj=1540
[2023-Aug-05 20:37:42.510329] DEBUG global - Count: 0; base count: 13
[2023-Aug-05 20:37:42.510333] DEBUG global - Stack has been destroyed.
[2023-Aug-05 20:37:42.510340] DEBUG global - Calculating casualties for unnamed battle stack [6]: 0 of creatures.core.royalGriffin.name.plural from slot 1 of armyobj=1540
[2023-Aug-05 20:37:42.510345] DEBUG global - Count: 0; base count: 12
[2023-Aug-05 20:37:42.510349] DEBUG global - Stack has been destroyed.
[2023-Aug-05 20:37:42.510578] DEBUG global - removing object id=1540; address=1036a90c0; name=Грифоны
[2023-Aug-05 20:37:42.511524] DEBUG global - Hero hero.core.nagash.name visit ends.

[2023-Aug-05 20:37:42.528859] ERROR network - Network error receiving a pack. Connection has been closed
[2023-Aug-05 20:37:42.529146] INFO network - Thread listening for Connection with VCMI 1.3.0 (server) (ID: 1 UUID: e0c00824-0e0f-4658-866e-69ddd718c90f) ended
```

### `VCMI_client.txt`
```
[2023-Aug-05 20:37:38.410930] DEBUG animation - CReverseAnimation::init: stack Лорды вампиров
[2023-Aug-05 20:37:38.410985] DEBUG animation - CReverseAnimation::init: stack Грифоны
[2023-Aug-05 20:37:38.499460] DEBUG animation - Created MeleeAttackAnimation for Лорды вампиров
[2023-Aug-05 20:37:38.570102] DEBUG animation - Created HittedAnimation for Грифоны
[2023-Aug-05 20:37:38.695462] DEBUG animation - Created ReverseAnimation for Грифоны
[2023-Aug-05 20:37:38.695511] DEBUG animation - Created ReverseAnimation for Лорды вампиров
[2023-Aug-05 20:37:38.710836] DEBUG animation - CReverseAnimation::init: stack Грифоны
[2023-Aug-05 20:37:38.710892] DEBUG animation - CReverseAnimation::init: stack Лорды вампиров
[2023-Aug-05 20:37:38.801306] DEBUG animation - Activating next stack
[2023-Aug-05 20:37:40.915501] DEBUG animation - Created MovementStartAnimation for Черные рыцари
[2023-Aug-05 20:37:40.929518] DEBUG animation - CMovementStartAnimation::init: stack Черные рыцари
[2023-Aug-05 20:37:40.947663] DEBUG animation - Created MovementAnimation for Черные рыцари
[2023-Aug-05 20:37:40.947712] DEBUG animation - CMovementAnimation::init: stack Черные рыцари moves {BattleHex: x '3', y '8', hex '139'} -> {BattleHex: x '4', y '7', hex '123'}
[2023-Aug-05 20:37:41.034538] DEBUG animation - CMovementAnimation::init: stack Черные рыцари moves {BattleHex: x '4', y '7', hex '123'} -> {BattleHex: x '5', y '7', hex '124'}
[2023-Aug-05 20:37:41.122191] DEBUG animation - Created MovementEndAnimation for Черные рыцари
[2023-Aug-05 20:37:41.122237] DEBUG animation - CMovementEndAnimation::init: stack Черные рыцари
[2023-Aug-05 20:37:41.160337] DEBUG animation - Created MeleeAttackAnimation for Черные рыцари
[2023-Aug-05 20:37:41.228336] DEBUG animation - Created DeathAnimation for Грифоны
[2023-Aug-05 20:37:41.402022] DEBUG global - totalRedraw requested!
[2023-Aug-05 20:37:42.510424] DEBUG global - totalRedraw requested!
```

**To Reproduce**
Obviously, the bug appears from time to time. Judging by the call stack trace, the bug is related to getting a list of bonuses.

**Expected behavior**
The game continues its story as neither victory nor defeat is achieved.

**Actual behavior**
Namespace SIGNAL, Code 11 Segmentation fault: 11

**Did it work earlier?**
Yes

**Screenshots**
N/A

**Version**
 - OS: macOS 13.5
 - VCMI 1.3.0

**Additional context**
N/A

**Comment by IvanSavenko** [comment]
- Is this hero vs hero battle or hero vs monster?
- Were you the one to win the battle?
- Did all your creatures survived the battle or some stacks have died?

**Comment by gestern** [comment]
@IvanSavenko Hello, Ivan

- It was a battle with the computer (but apparently the problem is relevant to any type of battle). The rarity of occurrence is related to the likelihood of stack overwriting.
- I had to win the battle, so I got very upset. :)
- Survived, there were no losses

**Comment by IvanSavenko** [comment]
Understood, will look into it. 
Thanks for stack traces.

**Comment by IvanSavenko** [comment]
Very likely to be fixed in #2554
